src/classkit.cs

```diff
@@ -0,0 +1,336 @@
+	[BaseType (typeof (CLSObject))]
+	[DisableDefaultCtor]
+	interface CLSActivityItem {
+
+		[Export ("title")]
+		string Title { get; set; }
+
+		[Export ("identifier")]
+		string Identifier { get; }
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,4)]
+	[BaseType (typeof (CLSActivityItem))]
+	[DisableDefaultCtor]
+	interface CLSBinaryItem {
+
+		[Export ("value")]
+		bool Value { get; set; }
+
+		[Export ("valueType")]
+		CLSBinaryValueType ValueType { get; }
+
+		[Export ("initWithIdentifier:title:type:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (string identifier, string title, CLSBinaryValueType valueType);
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,4)]
+	[BaseType (typeof (CLSObject))]
+	[DisableDefaultCtor]
+	interface CLSContext {
+
+		[Export ("identifier")]
+		string Identifier { get; }
+
+		[NullAllowed, Export ("universalLinkURL", ArgumentSemantic.Assign)]
+		NSUrl UniversalLinkUrl { get; set; }
+
+		[Export ("type")]
+		CLSContextType Type { get; }
+
+		[Export ("title")]
+		string Title { get; set; }
+
+		[Export ("displayOrder")]
+		nint DisplayOrder { get; set; }
+
+		[Protected]
+		[NullAllowed, Export ("topic")]
+		NSString WeakTopic { get; set; }
+
+		[Export ("initWithType:identifier:title:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (CLSContextType type, string identifier, string title);
+
+		[Export ("active")]
+		bool Active { [Bind ("isActive")] get; set; }
+
+		[Export ("becomeActive")]
+		void BecomeActive ();
+
+		[Export ("resignActive")]
+		void ResignActive ();
+
+		// From CLSContext (Hierarchy) Category
+
+		[NullAllowed, Export ("parent")]
+		CLSContext Parent { get; }
+
+		[Export ("removeFromParent")]
+		void RemoveFromParent ();
+
+		[Export ("addChildContext:")]
+		void AddChild (CLSContext childContext);
+
+		[Async]
+		[Export ("descendantMatchingIdentifierPath:completion:")]
+		void FindDescendantMatching (string [] identifierPath, Action<CLSContext, NSError> completion);
+
+		// From CLSContext (Activity) Category
+
+		[NullAllowed, Export ("currentActivity")]
+		CLSActivity CurrentActivity { get; }
+
+		[Export ("createNewActivity")]
+		CLSActivity CreateNewActivity ();
+	}
+
+	interface ICLSDataStoreDelegate { }
+
+	[NoWatch, NoTV, NoMac, iOS (11,4)]
+	[Protocol, Model]
+	[BaseType (typeof (NSObject))]
+	interface CLSDataStoreDelegate {
+
+		[Abstract]
+		[Export ("createContextForIdentifier:parentContext:parentIdentifierPath:")]
+		CLSContext CreateContext (string identifier, CLSContext parentContext, string [] parentIdentifierPath);
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,4)]
+	[BaseType (typeof (NSObject))]
+	[DisableDefaultCtor]
+	interface CLSDataStore {
+
+		[Static]
+		[Export ("shared")]
+		CLSDataStore Shared { get; }
+
+		[Export ("mainAppContext")]
+		CLSContext MainAppContext { get; }
+
+		[NullAllowed, Export ("activeContext")]
+		CLSContext ActiveContext { get; }
+
+		[NullAllowed, Export ("runningActivity")]
+		CLSActivity RunningActivity { get; }
+
+		[Wrap ("WeakDelegate"), Protocolize]
```

Comment:
`[Protocolize]` was to keep compatibility with classic.

Reply:
@spouliot Yeah I know but something weird is happening

If I do 

```csharp
[Wrap ("WeakDelegate")]
[NullAllowed]
CLSDataStoreDelegate Delegate { get; set; }

[NullAllowed, Export ("delegate", ArgumentSemantic.Weak)]
NSObject WeakDelegate { get; set; }
```

I get a warning `BI1110: bgen: The property ClassKit.CLSDataStore.Delegate exposes a model (ClassKit.CLSDataStoreDelegate). Please expose the corresponding protocol type instead (ClassKit.ICLSDataStoreDelegate).` which is completely expected and correct, if I change the code to 

```csharp
[Wrap ("WeakDelegate")]
[NullAllowed]
ICLSDataStoreDelegate Delegate { get; set; }

[NullAllowed, Export ("delegate", ArgumentSemantic.Weak)]
NSObject WeakDelegate { get; set; }
```

I get a build error

```
xamarin-macios/src/build/ios/native/ClassKit/CLSDataStore.g.cs(215,20): error CS0266: Cannot implicitly convert type `ClassKit.ICLSDataStoreDelegate' to `Foundation.NSObject'. An explicit conversion exists (are you missing a cast?)
xamarin-macios/src/build/ios/native/ClassKit/CLSDataStore.g.cs(215,20): error CS0266: Cannot implicitly convert type `ClassKit.ICLSDataStoreDelegate' to `Foundation.NSObject'. An explicit conversion exists (are you missing a cast?)
```

Generated code looks like this:

```csharp
[CompilerGenerated]
public ICLSDataStoreDelegate Delegate {
	get {
		return WeakDelegate as ICLSDataStoreDelegate;
	}
	set {
		WeakDelegate = value;
	}
}
```

The only way to get the right code is using `[Protocolize]` 😞 

```csharp
[CompilerGenerated]
public ICLSDataStoreDelegate Delegate {
	get {
		return WeakDelegate as ICLSDataStoreDelegate;
	}
	set {
		var rvalue = value as NSObject;
		if (value != null && rvalue == null)
			throw new ArgumentException ("The object passed of type " + value.GetType () + " does not derive from NSObject");
		WeakDelegate = rvalue;
	}
}
```

Or what am I missing here?

Reply:
This might need a generator change 😒

We didn't used to add the `WeakDelegate` property for Delegate properties that were typed as protocols. We have a few, but they're all readonly from what I can see, and don't run into this problem.

I think it's fine to comment out the WeakDelegate property, and file an issue for the generator.

Reply:
@spouliot filed as https://github.com/xamarin/xamarin-macios/issues/3869